stdlib/src/os/path/path.mojo

```diff
@@ -390,3 +391,156 @@ def split[PathLike: os.PathLike, //](path: PathLike) -> (String, String):
+        ord("-"),
+        ord("0"),
+        ord("1"),
+        ord("2"),
+        ord("3"),
+        ord("4"),
+        ord("5"),
+        ord("6"),
+        ord("7"),
+        ord("8"),
+        ord("9"),
+    )
+    if int(byte) in shell_variables:
+        return True
+    return False
+
+
+fn _is_alphanumeric(byte: Byte) -> Bool:
+    """Checks if `byte` is an ASCII letter, number, or underscore.
+
+    Args:
+        byte: The byte to check.
+
+    Returns:
+        True if the byte is an ASCII letter, number, or underscore and False otherwise.
+    """
+    var b = int(byte)
+    return (
+        b == ord("_")
+        or ord("0") <= b
+        and b <= ord("9")
+        or ord("a") <= b
+        and b <= ord("z")
+        or ord("A") <= b
+        and b <= ord("Z")
+    )
+
+
+fn _parse_variable_name(bytes: Span[Byte]) -> Tuple[String, Int]:
+    """Returns the environment variable name and the byte count required to extract it.
+    For `${}` expansions, two additional bytes are added to the byte count to account for the braces.
+
+    Args:
+        bytes: The bytes to extract the environment variable name from.
+
+    Returns:
+        The environment variable name and the byte count required to extract it.
+    """
+    if bytes[0] == ord("{"):
+        if (
+            len(bytes) > 2
+            and _is_shell_special_variable(bytes[1])
+            and bytes[2] == ord("}")
+        ):
+            return String(StringSlice(unsafe_from_utf8=bytes[1:2])), 3
+
+        # Scan until the closing brace or the end of the bytes.
+        var i = 1
+        while i < len(bytes):
+            if bytes[i] == ord("}"):
+                if i == 1:
+                    return String("${}"), 2
+                return String(StringSlice(unsafe_from_utf8=bytes[1:i])), i + 1
+            i += 1
+        return String("${"), 1
+    elif _is_shell_special_variable(bytes[0]):
+        return String(StringSlice(unsafe_from_utf8=bytes[0:1])), 1
+
+    # Scan until we hit an invalid character in environment variable names.
+    var i = 0
+    while i < len(bytes) and _is_alphanumeric(bytes[i]):
+        i += 1
+
+    return String(StringSlice(unsafe_from_utf8=bytes[:i])), i
+
+
+fn expandvars[PathLike: os.PathLike, //](path: PathLike) -> String:
+    """Replaces `${var}` or `$var` in the path with values from the current environment variables.
+    Undefined variables should be left alone.
+
+    Parameters:
+        PathLike: The type conforming to the os.PathLike trait.
+
+    Args:
+        path: The path to expand.
+
+    Returns:
+        The input path with environment variables expanded.
+    """
+    var path_str = path.__fspath__()
+    var bytes = path_str.as_bytes()
+    var buf = String()
+
+    # Byte scanning should be fine, ${} is ASCII.
+    i = 0
+    j = 0
+    while j < len(bytes):
+        if bytes[j] == ord("$") and j + 1 < len(bytes):
+            if not buf:
+                buf._buffer.reserve(new_capacity=2 * len(bytes))
```

Comment:
**Suggestion** This sort of suggests the public API of `String` ought to have a "reserve" like functionality to avoid needing to access the internal `_buffer` data member.  Do you mind filing a GitHub issue for this?

Reply:
I initially started out with a `List[Byte]` buffer and constructed a `String` from that at the end. But, writing directly to a `String` seemed more straightforward to understand. I think my need here would be solved with some sort of StringBuilder or being able to call `write` or `write_bytes` on `List[Byte]`. It could be handy to be able to allocate capacity ahead of time, I'll open up a Github issue and see the consensus is.

In the meantime, should I avoid using the internal `_buffer` for the time being?

Reply:
Thanks for the GH issue!  You can keep using `_buffer` for now as there are other users we'd need to fix up in the future too.  Appreciate it.